aws-api/src/main/java/com/amplifyframework/api/aws/GsonListDeserializer.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.api.aws;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Custom list deserializer since some lists come back not as an array of the items but as an object which contains
+ * an items property with the list of items and a nextToken property for pagination purposes.
+ */
+final class GsonListDeserializer implements JsonDeserializer<List<Object>> {
+    private final Gson singleGson;
+    private final Gson listGson;
+
+    GsonListDeserializer() {
+        this.singleGson = GsonFactory.create();
+        this.listGson = GsonFactory.create(Collections.singletonMap(List.class, this));
```

Comment:
Injecting these into the constructor is a pretty good approach, but I think a better is to use the `JsonDeserializationContext` that is already passed in.

Instead of:

```
listGson.fromJson(item, clazz)
```

use:

```
context.deserialize(item, clazz)
```

And instead of:

```
singleGson.fromJson(json, typeOfT)
```

use (you have to iterate through the array manually rather than just deserializing `json` itself to avoid endless recursion):

```
List<Object> items = new ArrayList();
for(JsonElement item : json.getAsJsonArray()) {
    items.add(context.deserialize(item, clazz);
}
return items;
```

Two main benefits to this approach:
 - No dependencies at all
 - You are guaranteed to be using the same Gson instance that was provided to deserialize the root node.

Reply:
Awesome!! Thanks. (We implemented this together on call. Pushing update now.)